localstack/testing/pytest/fixtures.py

```diff
@@ -1131,6 +1131,154 @@ def _check_logs(func_name: str, expected_lines: List[str] = None):
+        attached_policies = iam_client.list_attached_user_policies(UserName=username)[
+            "AttachedPolicies"
+        ]
+        for attached_policy in attached_policies:
+            try:
+                iam_client.detach_user_policy(
+                    UserName=username, PolicyArn=attached_policy["PolicyArn"]
+                )
+            except Exception:
+                LOG.debug(
+                    "Error detaching policy '%s' from user '%s'",
+                    attached_policy["PolicyArn"],
+                    username,
+                )
+        try:
+            iam_client.delete_user(UserName=username)
+        except Exception:
+            LOG.debug("Error deleting user '%s' during test cleanup", username)
+
+
+@pytest.fixture
+def wait_and_assume_role(sts_client):
+    def _wait_and_assume_role(role_arn: str, session_name: str = None):
+        if not session_name:
+            session_name = f"session-{short_uid()}"
+
+        def assume_role():
+            return sts_client.assume_role(RoleArn=role_arn, RoleSessionName=session_name)[
+                "Credentials"
+            ]
+
+        # need to retry a couple of times before we are allowed to assume this role in AWS
+        keys = retry(assume_role, sleep=5, retries=4)
+        return keys
+
+    return _wait_and_assume_role
+
+
+@pytest.fixture
+def create_role(iam_client):
+    role_names = []
+
+    def _create_role(**kwargs):
+        if not kwargs.get("RoleName"):
+            kwargs["RoleName"] = f"role-{short_uid()}"
+        result = iam_client.create_role(**kwargs)
+        role_names.append(result["Role"]["RoleName"])
+        return result
+
+    yield _create_role
+
+    for role_name in role_names:
+        # detach policies
+        attached_policies = iam_client.list_attached_role_policies(RoleName=role_name)[
+            "AttachedPolicies"
+        ]
+        for attached_policy in attached_policies:
+            try:
+                iam_client.detach_role_policy(
+                    RoleName=role_name, PolicyArn=attached_policy["PolicyArn"]
+                )
+            except Exception:
+                LOG.debug(
+                    "Could not detach role policy '%s' from '%s' during cleanup",
+                    attached_policy["PolicyArn"],
+                    role_name,
+                )
+        role_policies = iam_client.list_role_policies(RoleName=role_name)["PolicyNames"]
+        for role_policy in role_policies:
+            try:
+                iam_client.delete_role_policy(RoleName=role_name, PolicyName=role_policy)
+            except Exception:
+                LOG.debug(
+                    "Could not delete role policy '%s' from '%s' during cleanup",
+                    role_policy,
+                    role_name,
+                )
+        try:
+            iam_client.delete_role(RoleName=role_name)
+        except Exception:
+            LOG.debug("Could not delete role '%s' during cleanup", role_name)
```

Comment:
Not super happy about how we do the cleanup for these things. Especially when we have fixtures for a few of the dependent/nested resources. But definitely out of scope here.

Reply:
Yes, me neither. During IAM testing I introduced that, since the AWS account was spammed full with roles, which never got deleted, due to attached/inline policies not deleted yet.